### PR TITLE
Feat l1 l2 exchanger

### PR DESCRIPTION
**Pull request by dgornjakovic**

### CODE REVIEW

File: contracts/Token/OpThales_L1.sol [hunk 1]
```diff
@@ -1,16 +1,42 @@
...
+import {iOVM_L1StandardBridge} from "@eth-optimism/contracts/iOVM/bridge/tokens/iOVM_L1StandardBridge.sol";
 
+contract OpThales is ERC20, Ownable {
+    string public __name = "Optimistic Thales Token";
+    string public __symbol = "opTHALES";
+    uint8 public constant __decimals = 18;
     uint private INITIAL_TOTAL_SUPPLY = 100000000;
 
-    constructor () public ERC20("Optimistic Thales Token", "opTHALES") {
+    event NameChanged(string name);
```
Comment (dgornjakovic) [inline contracts/Token/OpThales_L1.sol:14]: shouldnt events be listed at the bottom at the contract per convention?

File: contracts/Token/OpThales_L1.sol [hunk 2]
```diff
@@ -1,16 +1,42 @@
...
+    event SymbolChanged(string symbol);
+
+    function name() public view override returns (string memory) {
```
Comment (dgornjakovic) [inline contracts/Token/OpThales_L1.sol:17]: if this is the getter, then __name should be private

File: contracts/Token/OpThales_L1.sol [hunk 3]
```diff
@@ -1,16 +1,42 @@
...
+        return __name;
+    }
+
+    function symbol() public view override returns (string memory) {
```
Comment (dgornjakovic) [inline contracts/Token/OpThales_L1.sol:21]: if this is the getter then __symbol should be private

File: contracts/Token/OpThales_L2.sol [hunk 4]
```diff
@@ -36,6 +43,18 @@ contract OpThales is IL2StandardERC20, ERC20 {
     ) public ERC20(_name, _symbol) {
         l1Token = _l1Token;
```
Comment (dgornjakovic) [inline contracts/Token/OpThales_L2.sol:44]: whats the point of default values in lines 9 and 10 if it has to be passed through constructor?

File: contracts/exchange_L1_L2/ThalesExchanger.sol [hunk 5]
```diff
@@ -36,8 +40,10 @@ contract ThalesExchanger is IThalesExchanger, Owned, ReentrancyGuard, Pausable {
         ThalesToken = IERC20(thalesAddress);
         OpThalesToken = IERC20(opThalesAddress);
         L1Bridge = iOVM_L1ERC20Bridge(_l1BridgeAddress);
-        OpThalesToken.approve(_l1BridgeAddress, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff);
+        OpThalesToken.approve(_l1BridgeAddress, MAX_APPROVAL);
         l2TokenAddress = _l2TokenAddress;
+        enabledThalesToOpThales = true;
```
Comment (dgornjakovic) [inline contracts/exchange_L1_L2/ThalesExchanger.sol:45]: set this as inline initialization in lines 25 and 26

File: contracts/exchange_L1_L2/ThalesExchanger.sol [hunk 6]
```diff
@@ -51,38 +57,55 @@ contract ThalesExchanger is IThalesExchanger, Owned, ReentrancyGuard, Pausable {
...
         L1Bridge = iOVM_L1ERC20Bridge(_l1BridgeAddress);
-        OpThalesToken.approve(_l1BridgeAddress, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff);
+        OpThalesToken.approve(_l1BridgeAddress, MAX_APPROVAL);
         emit L1BridgeChanged(_l1BridgeAddress);
     }
 
     function exchangeThalesToOpThales(uint amount) external nonReentrant notPaused {
         require(OpThalesToken.balanceOf(address(this)) >= amount, "Insufficient Exchanger OpThales funds");
         require(ThalesToken.allowance(msg.sender, address(this)) >= amount, "No allowance");
+        require(enabledThalesToOpThales, "Exchanging disabled");
```
Comment (dgornjakovic) [inline contracts/exchange_L1_L2/ThalesExchanger.sol:81]: should be the first check imo

File: contracts/Token/OpThales_L2.sol (hunk 4 above)
Comment (kirilantevski) [reply contracts/Token/OpThales_L2.sol:44]: No point at all

File: contracts/Token/OpThales_L1.sol (hunk 2 above)
Comment (kirilantevski) [reply contracts/Token/OpThales_L1.sol:17]: we have to override the getters from ERC-20
Yes, I can do it.

File: contracts/exchange_L1_L2/ThalesExchanger.sol (hunk 5 above)
Comment (kirilantevski) [reply contracts/exchange_L1_L2/ThalesExchanger.sol:45]: ok

File: contracts/Token/OpThales_L1.sol (hunk 1 above)
Comment (kirilantevski) [reply contracts/Token/OpThales_L1.sol:14]: Yes, let's proceed that way

File: contracts/Token/OpThales_L1.sol (hunk 3 above)
Comment (kirilantevski) [reply contracts/Token/OpThales_L1.sol:21]: OK

File: contracts/exchange_L1_L2/ThalesExchanger.sol (hunk 6 above)
Comment (kirilantevski) [reply contracts/exchange_L1_L2/ThalesExchanger.sol:81]: OK

**Pull request merged**